rfcs/text/0009_screenshot_mode_service.md

```diff
@@ -0,0 +1,151 @@
+- Start Date: 2020-03-02
+- RFC PR: (leave this empty)
+- Kibana Issue: (leave this empty)
+
+# Summary
+
+Currently, the applications that support screenshot reports are:
+ - Dashboard
+ - Visualize Editor
+ - Canvas
+
+Kibana UI code should be aware when the page is rendering for the purpose of
+capturing a screenshot. There should be a service to interact with low-level
+code for providing that awareness. Reporting would interact with this service
+to improve the quality of the Kibana Reporting feature for a few reasons:
+
+ - Fewer objects in the headless browser memory since interactive code doesn't run
+ - Fewer API requests made by the headless browser for features that don't apply in a non-interactive context
+
+**Screenshot mode service**
+
+The Reporting-enabled applications should use the recommended practice of
+having a customized URL for Reporting. The customized URL renders without UI
+features like navigation, auto-complete, and anything else that wouldn't make
+sense for non-interactive pages.
+
+However, applications are one piece of the UI code in a browser, and they have
+dependencies on other UI plugins. Apps can't control plugins and other things
+that Kibana loads in the browser.
+
+This RFC proposes a Screenshot Mode Service as a low-level plugin that allows
+other plugins (UI code) to make choices when the page is rendering for a screenshot.
+
+More background on how Reporting currently works, including the lifecycle of
+creating a PNG report, is here: https://github.com/elastic/kibana/issues/59396
+
+# Motivation
+
+The Reporting team wants all applications to support a customized URLs, such as
+Canvas does with its `#/export/workpad/pdf/{workpadId}` UI route. The
+customized URL is where an app can solve any rendering issue in a PDF or PNG,
+without needing extra CSS to be injected into the page.
+
+However, many low-level plugins have been added to the UI over time. These run
+on every page and an application can not turn them off. Reporting performance
+is negatively affected by this type of code. When the Reporting team analyzes
+customer logs to figure out why a job timed out, we sometimes see requests for
+the newsfeed API and telemetry API: services that aren't needed during a
+reporting job.
+
+In 7.12.0, using the customized `/export/workpad/pdf` in Canvas, the Sample
+Data Flights workpad loads 163 requests. Most of thees requests don't come from
```

Comment:
> loads 163 requests.

And most of them performed to fetch plugin bundles?
Doesn't Headless browser support caching? `puppeteer` supports it
https://pptr.dev/#?product=Puppeteer&version=v2.0.0&show=api-pagesetcacheenabledenabled
Can we use `Puppeteer` or borrow their caching implementation? 
If no, Core can consider serving a single bundle for all the Kiban plugins. We decided not to implement this logic when we've added long-term caching for bundle assets. So ideally, we would fix the caching problem for the headless browser.

Reply:
> And most of them performed to fetch plugin bundles?

That is correct. Since they are not fetched as background requests of plugins (e.g internal API requests for telemetry and newsfeed) there are limited options for a plugin to be able to reduce the 163 number.

> Doesn't Headless browser support caching? puppeteer supports it
https://pptr.dev/#?product=Puppeteer&version=v2.0.0&show=api-pagesetcacheenabledenabled
Can we use Puppeteer or borrow their caching implementation?

Caching is disabled since there is network request interception in-place in Kibana Reporting, to implement [the network policy rules](https://www.elastic.co/guide/en/kibana/current/reporting-network-policy.html), and prevent leaking credentials to 3rd parties. When network request interception is enabled, caching is disabled

> Can we use Puppeteer or borrow their caching implementation?

We are using Puppeteer today. 

NOTE: I just noticed this PR has been merged, which can offer help: https://github.com/puppeteer/puppeteer/pull/6996

Even if we are able to restore caching, it should be understood that the cache is cleared in between Reporting jobs. The browser is "run as" multiple different users so we do not want to use a single cache for multiple jobs.

> If no, Core can consider serving a single bundle for all the Kiban plugins. We decided not to implement this logic when we've added long-term caching for bundle assets. So ideally, we would fix the caching problem for the headless browser.

This would be fantastic for Reporting! 

Thanks for your comments!